### PR TITLE
Add RMSNorm support for BF16 and FP16 weights, add unit tests to verify BF16, FP16 weight results for fwd, bwd

### PR DESCRIPTION
**Pull request by lessw2020**
Based on issue (https://github.com/Dao-AILab/quack/issues/12), this PR adds:
1 - Support for RMSNorm weights in BF16, FP16, or FP32.  Previous code was limited to weights only in Fp32. 
Note that per @tridao, I've implemented it such that partial weight gradients are always computed in fp32. 
("we still want to compute the partial weight gradient in fp32 before summing up, then convert that to bf16")

* Now running in TorchTitan with FSDP2 and successfully training! 
* Speedup on LLama3-8B, B200*8, FSDP2 with TorchTitan:
Screenshot 2025-07-15 at 4 06 37 PM

2 - adds unit tests to verify BF16 and FP16 weights produce matching results to reference, for both forward and backwards. 
*   Forward pass tests:
    *   test_rmsnorm_bf16_weights 
    *   test_rmsnorm_fp16_weights 
*   Backward pass tests:
    *   test_rmsnorm_bf16_weights_backward 
    *   test_rmsnorm_fp16_weights_backward. 
    
 2a - update input validation test to verify fp16, bf16 and fp32 are accepted.  
 use fp64 input to test bad weight input. 
    
3 - compile cache is also updated to key on weight dtype as well. 

Testing:
a - verified all previous unit tests passing as before (337 passing)
b - add and run new unit tests (test_rmsnorm_bf16_weights is passing.
Screenshot 2025-07-15 at 11 05 18 AM
Screenshot 2025-07-15 at 12 22 42 PM

In TorchTitan:

#### =====================================================
Median Tokens/Second (excluding step 1): 9854.0
Max Memory Usage: 80.17 GiB

####  =====================================================
Median Tokens/Second (excluding step 1): 8752.0
Max Memory Usage: 90.37 GiB

### CODE REVIEW

File: quack/rmsnorm.py [hunk 1]
```diff
@@ -773,3 +809,39 @@ def rmsnorm(x: torch.Tensor, weight: torch.Tensor, eps: float = 1e-6) -> torch.T
         Normalized output tensor of same shape as x
     """
     return RMSNormFunction.apply(x, weight, eps)
+
+
+class QuackRMSNorm(torch.nn.Module):
```
Comment (tridao) [inline quack/rmsnorm.py:814]: we can just subclass torch.nn.RMSNorm instead? and only change the forward?

File: quack/rmsnorm.py [hunk 2]
```diff
@@ -632,9 +642,22 @@ def kernel(
                     tXsdW_other = cute.make_tensor(tXsdW.iterator + i * sdW.stride[0], tXsdW.layout)
                     cute.autovec_copy(tXsdW_other, tXrdW_other)
                     tXrdW.store(tXrdW.load() + tXrdW_other.load())
-                cute.copy(copy_atom_store_dW, tdWrdW, tdWgdW, pred=tdWpdW)
+                # Only convert for storing to weight.dtype if weights are not fp32
```
Comment (tridao) [inline quack/rmsnorm.py:645]: `tdWgdW.element_type` should always be fp32, since we allocate dw_partial to be in fp32. 
You can just put an assert here that type is fp32.

**Comment by tridao** [comment]
Wow amazing that you're already seeing speedup in torch titan!

File: quack/rmsnorm.py (hunk 1 above)
Comment (lessw2020) [reply quack/rmsnorm.py:803]: in theory yes, but I started trying this tonight and hit various cute dsl issues in the process.
Thus, I'd like to push this off as a seperate PR rather than try and do it here as we already have a lot of changes with bf16/fp16 support coming in. 
subclassing may also give us free torch.compile support as that's a different problem atm (we have to provide a dummy wrapper for compile to not try and dive into the quack code). 
Opened Issue https://github.com/Dao-AILab/quack/issues/18 to track torch.compile issue (and by proxy, rmsnorm subclassing)

**Review by tridao — APPROVED** [review]
